### PR TITLE
incremental sync

### PR DESCRIPTION
**Pull request by tantaman**
- `apps/mobile/app/sync/SyncedExpoDB.ts` would become a core component in a `@vlcn.io` or `expo.dev` package. This class does the mapping from `expo-sqlite` bindings to the data needed by the sync layer.
- The server (server.js) and client (App.tsx) aren't changed too much.

Since `expo-sqlite` doesn't support `blob` or `bigint` bindings yet, I had to do a bunch of conversions of types to/from hex and back. Bigints I'm truncating to Number which... definitely isn't safe in the long run but works for now. The existing implementation is also truncating bigints.

The lack of `bigint` support does prevent us from using the auto-migration extension to update the client database on schema file changes.

The db schema file is currently duplicated between client and server. Is there a way to load a text file via import in Expo? E.g., in vite I'd do `import schema from 'path/to/schema.sql?raw'`. That way `apps/server/schemas/test.sql` can be shared between client and server.

# Example:

https://github.com/expo/todo-sync-example/assets/1009003/d84faf3a-fd03-4dcd-b3bb-efab08ae6ed7

This video also highlights a comedy of problems.

1. TinyBase re-writes the _entire database_ on every modification!
2. Each write done by TinyBase calls DB change listeners

Because of 1 & 2:
- We re-render the app once for every single row in the DB on every change!!
- We re-pull database changes for every single row in the DB on every mutation!
- Before this incremental update, we would send N copies of the _entire_ database to the backend on every single mutation.

### CODE REVIEW

**Comment by tantaman** [comment]
I'm not a Yarn user but are those massive diffs in a yarn lock file normal? Looks crazy.

**Comment by brentvatne** [comment]
> I'm not a Yarn user but are those massive diffs in a yarn lock file normal? Looks crazy.

yikes, that shouldn't happen. i believe this is likely due to using a more recent yarn version with node-linker, whereas i'm using yarn classic (due to react-native ecosystem quirks, hopefully soon to be resolved)

File: apps/mobile/app/db/init.ts [hunk 1]
```diff
@@ -1,20 +1,39 @@
...
+import { createSingletonDbProvider } from "../sync/SyncedExpoDB";
+import { cryb64 } from "@vlcn.io/ws-common";
+import {decode, encode} from 'base-64'
+if (!global.btoa) {
+  global.btoa = encode;
+}
+
+if (!global.atob) {
+  global.atob = decode;
+}
```
Comment (brentvatne) [inline apps/mobile/app/db/init.ts:11]: welcome to react-native 👀

File: apps/mobile/app/db/init.ts [hunk 2]
```diff
@@ -1,20 +1,39 @@
...
-export const db = SQLite.openDatabase("test.db");
+export const dbName = "test.db";
+export const db = SQLite.openDatabase(dbName);
+
+// TODO: ideally we can share the schema in a package between client and server
+// this is curently duplicated into /server/schemas/test.sql
+const schema = [
+  `CREATE TABLE IF NOT EXISTS "todo" ("id" PRIMARY KEY, "text", "completed" INTEGER DEFAULT 0);`,
+  `SELECT crsql_as_crr('todo');`
+];
```
Comment (brentvatne) [inline apps/mobile/app/db/init.ts:21]: we could create a `shared` directory that contains packages that we share between server and client, and add that to the workspaces list. i can follow up on that in another pr

File: apps/mobile/app/db/init.ts [hunk 3]
```diff
@@ -1,20 +1,39 @@
...
     false,
     () => {}
   );
 }
+
+export const dbProvider = createSingletonDbProvider({
+  dbName,
+  db,
+  // TODO: users shouldn't manually deal with any of this.
+  // The browser db wrappers of cr-sqlite support automigration
```
Comment (brentvatne) [inline apps/mobile/app/db/init.ts:35]: cc @alanjhughes - here's something we can follow up on supporting in expo-sqlite

**Comment by tantaman** [comment]
I also talked with @threepointone about using partykit as an ephemeral room with an in-memory DB to facilitate sync. Should be doable by using a WASM build of SQLite.

So we could do that if we want to keep the PartyKit integration rather than the custom websocket server being introduced here.

File: apps/mobile/App.tsx [hunk 4]
```diff
@@ -34,10 +34,27 @@ export default function App() {
 
 function TodoList() {
   const store = useStore();
-  const { syncEnabled, setSyncEnabled } = useSync();
+  // const { syncEnabled, setSyncEnabled } = useSync();
```
Comment (brentvatne) [inline apps/mobile/App.tsx:37]:
```suggestion
```

**Review by brentvatne — APPROVED** [review]

**Review by alanjhughes — APPROVED** [review]

**Comment by tantaman** [comment]
@alanjhughes  - note that you'll have to use the _exact_ package versions specified in the PR ("@vlcn.io/ws-client": "^0.1.0-next.5", "@vlcn.io/ws-server": "^0.1.0-next.2") as they're special back-ported versions to work with the build of cr-sqlite (v0.14.0) you have.

I guess those version specifiers should not include `^` but rather be:

```
"@vlcn.io/ws-client": "0.1.0-next.5",
"@vlcn.io/ws-server": "0.1.0-next.2"
```

**Comment by alanjhughes** [comment]
@tantaman - no problem. I'm away today but I'll be working on it again tomorrow

**Pull request merged**